README.md
# motor-control-iio-modules

This repository contains the Kernel drivers for the Kd240 Motor Control
Application, leveraging the Industrial IO (IIO) framework. The drivers
are built on target using DKMS (Dynamic Kernel Module Support) framework.
 They are packaged for debian installation as shown in the next section

# Steps to build and install dkms drivers on target

```
  $ sudo apt install dkms
  $ sudo rm -r /usr/src/motor-control-iio-modules-0.1
  $ sudo mkdir -p /usr/src/motor-control-iio-modules-0.1
  $ git clone  https://github.com/Xilinx/motor-control-iio-modules
  $ sudo cp motor-control-iio-modules/src/* /usr/src/motor-control-iio-modules-0.1/
  $ sudo cp motor-control-iio-modules/debian/motor-control-iio-modules.dkms
/usr/src/motor-control-iio-modules-0.1/dkms.conf
  $ sudo dkms add -m motor-control-iio-modules -v 0.1
  $ sudo dkms build -m motor-control-iio-modules -v 0.1
  $ sudo dkms install -m motor-control-iio-modules -v 0.1
```
* `xmutil loadapp kd240-motor-ctrl-qei` will load the dkms installed drivers on target

# Verify module installation

```
$ modinfo hls-pwm-gen
filename:       /lib/modules/5.15.0-9002-xilinx-zynqmp/updates/dkms/hls-pwm-gen.ko
author:         Vivekananda Dayananda <vivekananda.dayananda@amd.com>
license:        GPL
srcversion:     623691B9B83CA4AB91F8695
alias:          of:N*T*Cxlnx,hls-pwm-gen-1.0C*
alias:          of:N*T*Cxlnx,hls-pwm-gen-1.0
depends:
name:           hls_pwm_gen
vermagic:       5.15.0-9002-xilinx-zynqmp SMP mod_unload modversions aarch64
```

# License

(C) Copyright 2022 - 2023 Advanced Micro Devices, Inc.\
SPDX-License-Identifier: GPL-2.0


debian/changelog
motor-control-iio-modules (0.1-0xlnx1) jammy; urgency=medium

  * Initial release

 -- Vivekananda Dayananda <vivekananda.dayananda@amd.com>  Mon, 04 Dec 2023 13:24:35 -0800


src/hls-qei-axi.c
// SPDX-License-Identifier: GPL-2.0
/*
 * HLS QEI driver
 *
 * Copyright (C) 2023, Advanced Micro Devices, Inc.
 *
 * Description:
 * This driver is developed for HLS quadrature encoder. The driver supports INDIO Mode
 * and supports encoder data and direction monitoring via IIO sysfs interface, buffer capable.
 */

#include <linux/delay.h>
#include <linux/iio/buffer.h>
#include <linux/iio/iio.h>
#include <linux/iio/kfifo_buf.h>
#include <linux/iio/sysfs.h>
#include <linux/kthread.h>
#include <linux/module.h>
#include <linux/of_address.h>
#include <linux/platform_device.h>
#include <linux/sched/task.h>

/* register space */
#define AP_CTRL	0x00
#define REG_CPR_I	0x10
#define REG_CPR_O	0x18
#define REG_CTRL	0x20		/* register for direction clockwise or counterclockwise */
#define REG_THETA_RPM	0x28
#define REG_DIR	0x38
#define REG_ERR	0x48			/* clear on read COR */
#define MASK_UPPER	GENMASK(31, 16)
#define MASK_LOWER	GENMASK(15, 0)
#define DEFAULT_CPR	0x3E8
#define DEFAULT_AP_CTRL	0x00
#define SAMPLE_INTERVAL	0xFF
#define SIGNEX(v, sb) (((v) & (1 << ((sb) - 1))) ? ((v) | (~((1 << (sb)) - 1))) : (v))
#define COUNT_CHANNEL(num, _ext)				\
	{							\
	    .type = IIO_ROT,					\
	    .indexed = 1,					\
	    .channel = (num),					\
	    .address = REG_THETA_RPM,				\
	    .scan_index = (num),				\
	    .info_mask_separate = BIT(IIO_CHAN_INFO_RAW) |	\
				BIT(IIO_CHAN_INFO_SCALE),	\
	    .scan_type = {					\
			.sign = 'u',				\
			.realbits = 32,				\
			.storagebits = 32,			\
			.shift = 0,				\
			},					\
	    .extend_name = _ext,				\
	}							\

struct qei_data {
	void __iomem *base;
	u32 *data;
	u32 interval;
	struct task_struct *task;
};

static const struct iio_chan_spec rot_channels[] = {
	COUNT_CHANNEL(0, "rpm"),
	COUNT_CHANNEL(1, "theta_degree"),
	IIO_CHAN_SOFT_TIMESTAMP(2),

};

static inline void qei_read_reg(struct qei_data *st, unsigned int offset,
				u32 *data)
{
	*data = readl(st->base + offset);
}

static inline void qei_write_reg(struct qei_data *st, unsigned int offset,
				 u32 data)
{
	writel(data, st->base + offset);
}

/* function to convert raw value to degrees */
static int xqei_convert_to_deg(unsigned int *data)
{
	unsigned int tmp;
		tmp = MASK_UPPER & *data;
		tmp = tmp >> 16;
		tmp = (tmp * 360 / 1000);
		*data = tmp;
	return 0;
}

static int qei_read_attribute(struct iio_dev *indio_dev,
			      struct iio_chan_spec const *channels, int *val,
			      int *val2, long mask)
{
	struct qei_data *st = iio_priv(indio_dev);
	unsigned int data;

	switch (mask) {
	case IIO_CHAN_INFO_RAW:
		qei_read_reg(st, channels->address, &data);
		if (!(channels->channel)) {
			/* read RPM for channel 0 */
			*val = SIGNEX(data & MASK_LOWER, 16);
		} else {/* read theta for channel 1 */
			xqei_convert_to_deg(&data);
			*val = data;
		}
		return IIO_VAL_INT;

	case IIO_CHAN_INFO_SCALE:
		*val = 1;
		return IIO_VAL_INT;
	default:
		return -EINVAL;
	}
}

static int xqei_verify_write_success(struct iio_dev *indio_dev, unsigned int reg_in,
				     unsigned int reg_out, u32 data_in)
{
	struct qei_data *data = iio_priv(indio_dev);
	u32 data_out;

	qei_write_reg(data, reg_in, data_in);
	/* The HW requires 100ms to propagate data_in value to the shadow reg */
	mdelay(100);
	qei_read_reg(data, reg_out, &data_out);
	/* comparing shadow reg with data_in */
	if (data_out != data_in) {
		dev_warn(&indio_dev->dev, "value %u is out of bounds. Register [base + %u] failed to set\n",
			 data_in, reg_out);
		return -EINVAL;
	}
	return 0;
}

static ssize_t qei_ctrl_show(struct device *dev,
			     struct device_attribute *attr,
			     char *buf)
{
	struct iio_dev_attr *this_attr = to_iio_dev_attr(attr);
	struct iio_dev *indio_dev = dev_to_iio_dev(dev);
	struct qei_data *st = iio_priv(indio_dev);
	unsigned int val;
	int ret = 0;

	switch ((u32)this_attr->address) {
	case REG_CPR_I:
		qei_read_reg(st, REG_CPR_O, &val);
		ret = sprintf(buf, "%u\n", val);
		break;
	case REG_CTRL:
		qei_read_reg(st, REG_CTRL, &val);
		ret = sprintf(buf, "%u\n", val);
		break;
	case REG_DIR:
		qei_read_reg(st, REG_DIR, &val);
		ret = sprintf(buf, "%u\n", val);
		break;
	case REG_ERR:
		qei_read_reg(st, REG_ERR, &val);
		ret = sprintf(buf, "%u\n", val);
		break;
	case AP_CTRL:
		qei_read_reg(st, AP_CTRL, &val);
		ret = sprintf(buf, "%u\n", val);
		break;
	case SAMPLE_INTERVAL:
		val = st->interval;
		ret = sprintf(buf, "%u\n", val);
		break;
	default:
		return -EINVAL;
	}

	return ret;
}

static ssize_t qei_ctrl_store(struct device *dev,
			      struct device_attribute *attr,
			      const char *buf, size_t len)
{
	struct iio_dev_attr *this_attr = to_iio_dev_attr(attr);
	struct iio_dev *indio_dev = dev_to_iio_dev(dev);
	struct qei_data *st = iio_priv(indio_dev);
	unsigned long long data_in;
	int ret = 0;

	switch ((u32)this_attr->address) {
	case REG_CPR_I:
		ret = kstrtoull(buf, 10, &data_in);
		if (!ret)
			xqei_verify_write_success(indio_dev, REG_CPR_I, REG_CPR_O, data_in);
		else
			return ret;
		break;
	case REG_CTRL:
		ret = kstrtoull(buf, 10, &data_in);
		if (ret) {
			return ret;
		} else if (data_in == 1 || data_in == 0) {
			qei_write_reg(st, REG_CTRL, data_in);
		} else {
			dev_err(&indio_dev->dev, "1: 'A lead B', 0: 'B lead A'. Value %llu is out of range\n",
				data_in);
			return -EINVAL;
		}
		break;
	case REG_DIR:
		return -EINVAL;
	case REG_ERR:
		return -EINVAL;
	case AP_CTRL:
		ret = kstrtoull(buf, 10, &data_in);
		if (!ret)
			qei_write_reg(st, AP_CTRL, data_in);
		else
			return ret;
		break;
	case SAMPLE_INTERVAL:
		ret = kstrtoull(buf, 10, &data_in);
		if (!ret)
			st->interval = data_in;
		else
			return ret;
		break;
	default:
		return -EINVAL;
	}
return len;
}

static const struct of_device_id xqei_ids[] = {
	{ .compatible = "xlnx,hls-qei-axi-1.0", },
	{ /* sentinel */ }
};

MODULE_DEVICE_TABLE(of, xqei_ids);

static IIO_DEVICE_ATTR(error_status, 0644,
		       qei_ctrl_show,
		       qei_ctrl_store,
		       REG_ERR);

static IIO_DEVICE_ATTR(direction, 0644,
		       qei_ctrl_show,
		       qei_ctrl_store,
		       REG_DIR);

static IIO_DEVICE_ATTR(mode, 0644,
		       qei_ctrl_show,
		       qei_ctrl_store,
		       REG_CTRL);

static IIO_DEVICE_ATTR(cycles_per_revolution, 0644,
		       qei_ctrl_show,
		       qei_ctrl_store,
		       REG_CPR_I);

static IIO_DEVICE_ATTR(ap_ctrl, 0644,
		       qei_ctrl_show,
		       qei_ctrl_store,
		       AP_CTRL);

static IIO_DEVICE_ATTR(sample_interval_us, 0644,
		       qei_ctrl_show,
		       qei_ctrl_store,
		       SAMPLE_INTERVAL);

static struct attribute *qei_attributes[] = {
	&iio_dev_attr_cycles_per_revolution.dev_attr.attr,
	&iio_dev_attr_sample_interval_us.dev_attr.attr,
	&iio_dev_attr_error_status.dev_attr.attr,
	&iio_dev_attr_direction.dev_attr.attr,
	&iio_dev_attr_ap_ctrl.dev_attr.attr,
	&iio_dev_attr_mode.dev_attr.attr,
	NULL,
};

static const struct attribute_group qei_attribute_group = {
	.attrs = qei_attributes,
};

static const struct iio_info xqei_info = {
	.read_raw = &qei_read_attribute,
	.attrs = &qei_attribute_group,
};

static int xqei_fill_buffer(struct iio_dev *indio_dev)
{
	struct qei_data *st = iio_priv(indio_dev);
	/* data buffer for channel and time stamp */
	unsigned int data[1 + sizeof(s64) / sizeof(int)] = {0};
	unsigned int val, chan = 0, i = 0;
	s64 time;

	time = iio_get_time_ns(indio_dev);
	/*
	 * Read data for active channels from device
	 */
	for_each_set_bit(chan, indio_dev->active_scan_mask,
			 indio_dev->masklength) {
		qei_read_reg(st, REG_THETA_RPM, &val);
		if (!chan) {
			/* read RPM for channel 0 */
			data[i] = SIGNEX(val & MASK_LOWER, 16);
		} else {
			/* read theta for channel 1 */
			xqei_convert_to_deg(&val);
			data[i] = val;
		}
		i++;
	}

	iio_push_to_buffers_with_timestamp(indio_dev, data, time);

	return 0;
}

static int xqei_capture_thread(void *data)
{
	struct iio_dev *indio_dev = data;
	struct qei_data *st = iio_priv(indio_dev);
	u32 interval = st->interval;

	do {
		/* Read the data from sensor and push it to buffers */
		int ret = xqei_fill_buffer(indio_dev);

		if (ret < 0)
			return ret;
		usleep_range(interval, interval + 20);
	} while (!kthread_should_stop());

	return 0;
}

static int xqei_buffer_enable(struct iio_dev *indio_dev)
{
	struct qei_data *st = iio_priv(indio_dev);
	struct task_struct *task;

	task = kthread_create(xqei_capture_thread, (void *)indio_dev,
			      "%s", indio_dev->name);

	if (IS_ERR(task))
		return PTR_ERR(task);

	get_task_struct(task);
	wake_up_process(task);
	st->task = task;

	return 0;
}

static int xqei_buffer_disable(struct iio_dev *indio_dev)
{
	struct qei_data *st = iio_priv(indio_dev);

	if (st->task) {
		kthread_stop(st->task);
		put_task_struct(st->task);
		st->task = NULL;
	}
	return 0;
}

static const struct iio_buffer_setup_ops xqei_setup_ops = {
	.predisable = &xqei_buffer_disable,
	.postenable = &xqei_buffer_enable,
};

static int xqei_probe(struct platform_device *pdev)
{
	const struct of_device_id *id;
	struct iio_dev *indio_dev;
	struct qei_data *data;
	struct resource *res;
	int ret;

	if (!pdev->dev.of_node)
		return -ENODEV;

	id = of_match_node(xqei_ids, pdev->dev.of_node);

	if (!id)
		return -ENODEV;

	indio_dev = devm_iio_device_alloc(&pdev->dev, sizeof(*data));

	if (!indio_dev) {
		dev_err(&pdev->dev, "iio allocation failed!\n");
		return -ENOMEM;
	}

	data = iio_priv(indio_dev);
	indio_dev->dev.parent = &pdev->dev;
	indio_dev->info = &xqei_info;
	indio_dev->name = KBUILD_MODNAME;
	indio_dev->modes = INDIO_DIRECT_MODE;
	indio_dev->channels = rot_channels;
	indio_dev->num_channels = ARRAY_SIZE(rot_channels);
	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
	data->base = devm_ioremap_resource(&pdev->dev, res);
	data->interval = 500;

	if (IS_ERR(data->base))
		return PTR_ERR(data->base);

	qei_write_reg(data, REG_CPR_I, DEFAULT_CPR);
	qei_write_reg(data, AP_CTRL, DEFAULT_AP_CTRL);
	ret = devm_iio_kfifo_buffer_setup(&indio_dev->dev, indio_dev,
					  INDIO_BUFFER_SOFTWARE,
					  &xqei_setup_ops);
	if (ret)
		return ret;

	ret = iio_device_register(indio_dev);
	if (ret < 0)
		return ret;

	dev_info(&pdev->dev, "Successfully registered device XQEI encoder");
	platform_set_drvdata(pdev, indio_dev);

	return 0;
}

static int xqei_remove(struct platform_device *pdev)
{
	struct iio_dev *indio_dev = platform_get_drvdata(pdev);

	iio_device_unregister(indio_dev);
	return 0;
}

static struct platform_driver xqei_drv = {
	.probe = xqei_probe,
	.remove = xqei_remove,
	.driver = {
		.name = "hls-qei-axi",
		.of_match_table = of_match_ptr(xqei_ids),
	},
};
module_platform_driver(xqei_drv);

MODULE_LICENSE("GPL");
MODULE_AUTHOR("Vivekananda Dayananda <vivekananda.dayananda@amd.com>");


src/hls-pwm-gen.c
// SPDX-License-Identifier: GPL-2.0
/*
 * PWM GEN driver
 *
 * Copyright (C) 2023, Advanced Micro Devices, Inc.
 *
 * Description:
 * This driver is developed for Vitis xf_motorcontrol HLS library, IP PWM_GEN.
 * The driver supports INDIO Mode with control and monitoring of the IP via
 * IIO sysfs interface, also buffer capable.
 */

#include <linux/delay.h>
#include <linux/iio/buffer.h>
#include <linux/iio/iio.h>
#include <linux/iio/kfifo_buf.h>
#include <linux/iio/sysfs.h>
#include <linux/kthread.h>
#include <linux/module.h>
#include <linux/of_address.h>
#include <linux/platform_device.h>
#include <linux/sched/task.h>

/* register space */
#define AP_CTRL	0x00
#define PWM_FREQ	0x10
#define DEAD_CYCLES	0x18
#define PHASE_SHIFT	0x20
#define PWM_CYCLE	0x28	// debug reg
#define SAMPLE_II	0x38
#define DUTY_RATIO_A	0x40
#define DUTY_RATIO_B	0x50
#define DUTY_RATIO_C	0x60
/* defaults */
#define DEFAULT_PWM_FREQ	0x4E20
#define DEFAULT_DEAD_CYCLES	0xA
#define DEFAULT_PHASE_SHIFT	0x0
#define DEFAULT_SAMPLE_II	0x1
#define DEFAULT_AP_CTRL	0x1
#define MIN_DELAY_US	1000
#define MAX_DELAY_US	1200
/* channel spec */
#define VOLTAGE_CHANNEL(num, _ext)				\
	{							\
	    .type = IIO_VOLTAGE,				\
	    .indexed = 1,					\
	    .channel = (num),					\
	    .address = DUTY_RATIO_A + (num * 0x10),		\
	    .scan_index = (num),				\
	    .info_mask_separate = BIT(IIO_CHAN_INFO_RAW) |	\
				BIT(IIO_CHAN_INFO_SCALE),	\
	    .scan_type = {					\
			.sign = 'u',				\
			.realbits = 32,				\
			.storagebits = 32,			\
			.shift = 0,				\
			},					\
	    .extend_name = _ext,				\
	}							\

struct private_data {
	void __iomem *base;
	u32 *data;
	struct task_struct *task;
};

static const struct iio_chan_spec volt_duty_ratio_channels[] = {
	VOLTAGE_CHANNEL(0, "Va_duty_ratio"),
	VOLTAGE_CHANNEL(1, "Vb_duty_ratio"),
	VOLTAGE_CHANNEL(2, "Vc_duty_ratio"),
	IIO_CHAN_SOFT_TIMESTAMP(3),
};

static inline void pwmgen_read_reg(struct private_data *st, unsigned int offset,
				   u32 *data)
{
	*data = readl(st->base + offset);
}

static inline void pwmgen_write_reg(struct private_data *st, unsigned int offset,
				    u32 data)
{
	writel(data, st->base + offset);
}

static int svpmw_read_attribute(struct iio_dev *indio_dev,
				struct iio_chan_spec const *channels, int *val,
				int *val2, long mask)
{
	struct private_data *st = iio_priv(indio_dev);

	switch (mask) {
	case IIO_CHAN_INFO_RAW:
		pwmgen_read_reg(st, channels->address, val);
		*val2 = 16;
		return IIO_VAL_FRACTIONAL_LOG2;

	case IIO_CHAN_INFO_SCALE:
		*val = 1;
		*val2 = 0;
		return IIO_VAL_INT;
	default:
		return -EINVAL;
	}
	return -EINVAL;
}

/**
 * pwmgen_debugfs_reg_access - read or write register value
 * @indio_dev: IIO device structure
 * @reg: register offset
 * @writeval: value to write
 * @readval: value to read
 *
 * To read a value from the register:
 *   echo [pwmgen reg offset] > direct_reg_access
 *   cat direct_reg_access
 *
 * To write a value in a pwmgen register:
 *   echo [pwmgen_reg_offset] [value] > direct_reg_access
 */
static int pwmgen_debugfs_reg_access(struct iio_dev *indio_dev,
				     unsigned int reg, unsigned int writeval,
				     unsigned int *readval)
{
	struct private_data *pwmgen_data = iio_priv(indio_dev);
	struct device *dev = indio_dev->dev.parent;

	if (reg == 1) {
		pwmgen_read_reg(pwmgen_data, PWM_CYCLE, readval);
		dev_info(dev, "Value of debug register pwm_cycle with offset 0x%x is %d\n",
			 PWM_CYCLE, *readval);
	} else if (!readval) {
		pwmgen_write_reg(pwmgen_data, reg, writeval);
	} else {
		pwmgen_read_reg(pwmgen_data, reg, readval);
	}
	return 0;
}

static ssize_t pwmgen_ctrl_show(struct device *dev,
				struct device_attribute *attr,
				char *buf)
{
	struct iio_dev_attr *this_attr = to_iio_dev_attr(attr);
	struct iio_dev *indio_dev = dev_to_iio_dev(dev);
	struct private_data *st = iio_priv(indio_dev);
	unsigned int val;
	int ret = 0;

	switch ((u32)this_attr->address) {
	case PWM_FREQ:
		pwmgen_read_reg(st, PWM_FREQ, &val);
		ret = sprintf(buf, "%u\n", val);
		break;
	case DEAD_CYCLES:
		pwmgen_read_reg(st, DEAD_CYCLES, &val);
		ret = sprintf(buf, "%u\n", val);
		break;
	case PHASE_SHIFT:
		pwmgen_read_reg(st, PHASE_SHIFT, &val);
		ret = sprintf(buf, "%u\n", val);
		break;
	case SAMPLE_II:
		pwmgen_read_reg(st, SAMPLE_II, &val);
		ret = sprintf(buf, "%u\n", val);
		break;
	case AP_CTRL:
		pwmgen_read_reg(st, AP_CTRL, &val);
		ret = sprintf(buf, "%u\n", val);
		break;
	default:
		return -EINVAL;
	}

	return ret;
}

static ssize_t pwmgen_ctrl_store(struct device *dev,
				 struct device_attribute *attr,
				 const char *buf, size_t len)
{
	struct iio_dev_attr *this_attr = to_iio_dev_attr(attr);
	struct iio_dev *indio_dev = dev_to_iio_dev(dev);
	struct private_data *st = iio_priv(indio_dev);
	unsigned long long data_in;
	int ret = 0;

	switch ((u32)this_attr->address) {
	case PWM_FREQ:
		ret = kstrtoull(buf, 10, &data_in);
		if (!ret)
			pwmgen_write_reg(st, PWM_FREQ, data_in);
		else
			return ret;
		break;
	case DEAD_CYCLES:
		ret = kstrtoull(buf, 10, &data_in);
		if (!ret)
			pwmgen_write_reg(st, DEAD_CYCLES, data_in);
		else
			return ret;
		break;
	case PHASE_SHIFT:
		ret = kstrtoull(buf, 10, &data_in);
		if (!ret)
			pwmgen_write_reg(st, PHASE_SHIFT, data_in);
		else
			return ret;
		break;
	case SAMPLE_II:
		ret = kstrtoull(buf, 10, &data_in);
		if (!ret)
			pwmgen_write_reg(st, SAMPLE_II, data_in);
		else
			return ret;
		break;
	case AP_CTRL:
		ret = kstrtoull(buf, 10, &data_in);
		if (!ret)
			pwmgen_write_reg(st, AP_CTRL, data_in);
		else
			return ret;
		break;
	default:
		return -EINVAL;
	}
return len;
}

static IIO_DEVICE_ATTR(pwm_freq, 0644,
		       pwmgen_ctrl_show,
		       pwmgen_ctrl_store,
		       PWM_FREQ);

static IIO_DEVICE_ATTR(dead_cycles, 0644,
		       pwmgen_ctrl_show,
		       pwmgen_ctrl_store,
		       DEAD_CYCLES);

static IIO_DEVICE_ATTR(phase_shift, 0644,
		       pwmgen_ctrl_show,
		       pwmgen_ctrl_store,
		       PHASE_SHIFT);

static IIO_DEVICE_ATTR(sample_ii, 0644,
		       pwmgen_ctrl_show,
		       pwmgen_ctrl_store,
		       SAMPLE_II);

static IIO_DEVICE_ATTR(ap_ctrl, 0644,
		       pwmgen_ctrl_show,
		       pwmgen_ctrl_store,
		       AP_CTRL);

static struct attribute *pwmgen_attributes[] = {
	&iio_dev_attr_phase_shift.dev_attr.attr,
	&iio_dev_attr_dead_cycles.dev_attr.attr,
	&iio_dev_attr_sample_ii.dev_attr.attr,
	&iio_dev_attr_pwm_freq.dev_attr.attr,
	&iio_dev_attr_ap_ctrl.dev_attr.attr,
	NULL,
};

static const struct attribute_group pwmgen_attribute_group = {
	.attrs = pwmgen_attributes,
};

static const struct iio_info pwmgen_info = {
	.debugfs_reg_access = pwmgen_debugfs_reg_access,
	.read_raw = &svpmw_read_attribute,
	.attrs = &pwmgen_attribute_group,
};

static int pwmgen_fill_buffer(struct iio_dev *indio_dev)
{
	struct private_data *st = iio_priv(indio_dev);
	/* data buffer for channel data and timestamp */
	unsigned int data[3 + sizeof(s64) / sizeof(int)] = {0};
	unsigned int val, chan = 0, i = 0;
	s64 time;

	time = iio_get_time_ns(indio_dev);
	/*
	 * Read data for active channels from device
	 */
	for_each_set_bit(chan, indio_dev->active_scan_mask,
			 indio_dev->masklength) {
		pwmgen_read_reg(st, DUTY_RATIO_A + chan * 0x10, &val);
		data[i] = val;
		i++;
	}

	iio_push_to_buffers_with_timestamp(indio_dev, data, time);

	return 0;
}

static int pwmgen_capture_thread(void *data)
{
	struct iio_dev *indio_dev = data;

	do {
		/* Read the data from sensor and push it to buffers */
		int ret = pwmgen_fill_buffer(indio_dev);

		if (ret < 0)
			return ret;
		usleep_range(MIN_DELAY_US, MAX_DELAY_US);
	} while (!kthread_should_stop());

	return 0;
}

static int pwmgen_buffer_enable(struct iio_dev *indio_dev)
{
	struct private_data *st = iio_priv(indio_dev);
	struct task_struct *task;

	task = kthread_create(pwmgen_capture_thread, (void *)indio_dev,
			      "%s", indio_dev->name);

	if (IS_ERR(task))
		return PTR_ERR(task);

	get_task_struct(task);
	wake_up_process(task);
	st->task = task;

	return 0;
}

static int pwmgen_buffer_disable(struct iio_dev *indio_dev)
{
	struct private_data *st = iio_priv(indio_dev);

	if (st->task) {
		kthread_stop(st->task);
		put_task_struct(st->task);
		st->task = NULL;
	}
	return 0;
}

static const struct iio_buffer_setup_ops pwmgen_setup_ops = {
	.postenable = &pwmgen_buffer_enable,
	.predisable = &pwmgen_buffer_disable,
};

static const struct of_device_id pwmgen_ids[] = {
	{ .compatible = "xlnx,hls-pwm-gen-1.0", },
	{ /* sentinel */ }
};

MODULE_DEVICE_TABLE(of, pwmgen_ids);

static int pwmgen_probe(struct platform_device *pdev)
{
	const struct of_device_id *id;
	struct iio_dev *indio_dev;
	struct private_data *data;
	struct resource *res;
	int ret;

	if (!pdev->dev.of_node)
		return -ENODEV;

	id = of_match_node(pwmgen_ids, pdev->dev.of_node);

	if (!id)
		return -ENODEV;

	indio_dev = devm_iio_device_alloc(&pdev->dev, sizeof(*data));

	if (!indio_dev) {
		dev_err(&pdev->dev, "iio allocation failed!\n");
		return -ENOMEM;
	}

	data = iio_priv(indio_dev);
	indio_dev->dev.parent = &pdev->dev;
	indio_dev->info = &pwmgen_info;
	indio_dev->name = KBUILD_MODNAME;
	indio_dev->modes = INDIO_DIRECT_MODE;
	indio_dev->channels = volt_duty_ratio_channels;
	indio_dev->num_channels = ARRAY_SIZE(volt_duty_ratio_channels);
	res = platform_get_resource(pdev, IORESOURCE_MEM, 0);
	data->base = devm_ioremap_resource(&pdev->dev, res);

	if (IS_ERR(data->base))
		return PTR_ERR(data->base);

	pwmgen_write_reg(data, PWM_FREQ, DEFAULT_PWM_FREQ);
	pwmgen_write_reg(data, DEAD_CYCLES, DEFAULT_DEAD_CYCLES);
	pwmgen_write_reg(data, PHASE_SHIFT, DEFAULT_PHASE_SHIFT);
	pwmgen_write_reg(data, SAMPLE_II, DEFAULT_SAMPLE_II);
	pwmgen_write_reg(data, AP_CTRL, DEFAULT_AP_CTRL);
	ret = devm_iio_kfifo_buffer_setup(&indio_dev->dev, indio_dev,
					  INDIO_BUFFER_SOFTWARE,
					  &pwmgen_setup_ops);
	if (ret)
		return ret;

	ret = iio_device_register(indio_dev);
	if (ret < 0)
		return ret;

	dev_info(&pdev->dev, "Successfully registered device PWM-GEN encoder");
	platform_set_drvdata(pdev, indio_dev);

	return 0;
}

static int pwmgen_remove(struct platform_device *pdev)
{
	struct iio_dev *indio_dev = platform_get_drvdata(pdev);

	iio_device_unregister(indio_dev);
	return 0;
}

static struct platform_driver pwmgen_drv = {
	.probe = pwmgen_probe,
	.remove = pwmgen_remove,
	.driver = {
		.name = "hls-pwm-gen",
		.of_match_table = of_match_ptr(pwmgen_ids),
	},
};
module_platform_driver(pwmgen_drv);

MODULE_LICENSE("GPL");
MODULE_AUTHOR("Vivekananda Dayananda <vivekananda.dayananda@amd.com>");
